src/dropbox-base.js

```diff
@@ -265,10 +316,85 @@ export class DropboxBase {
             status: res.status,
           };
         }
+
+        if (data.refresh_token) {
+          return {
+            accessToken: data.access_token,
+            refreshToken: data.refresh_token,
+            accessTokenExpiresAt: getTokenExpiresAt(data.expires_in),
+          };
+        }
         return data.access_token;
       });
   }
 
+  /**
+   *
+   * @returns {Promise<[*, *]>}
+   */
+  checkAndRefreshAccessToken() {
+    const canRefresh = this.getRefreshToken() && this.getClientId() && this.getClientSecret();
+    const needsRefresh = this.getAccessTokenExpiresAt() &&
+            (new Date(Date.now() + TokenExpirationBuffer)) >= this.getAccessTokenExpiresAt();
+    const needsToken = !this.getAccessToken();
+    if ((needsRefresh || needsToken) && canRefresh) {
+      return this.refreshAccessToken();
+    }
+    return Promise.resolve();
+  }
+
+  /**
+   *
+   * @returns {Promise<[*, *]>}
+   */
+  refreshAccessToken() {
+    let refreshUrl = 'https://api.dropbox.com/oauth2/token';
+    const clientId = this.getClientId();
+    const clientSecret = this.getClientSecret();
+
+    if (!clientId) {
+      throw new Error('A client id is required. You can set the client id using .setClientId().');
+    }
+    if (!clientSecret) {
+      throw new Error('A client secret is required. You can set the client id using .setClientSecret().');
+    }
+    const headers = {};
+    headers['Content-Type'] = 'application/json';
+    refreshUrl += `?grant_type=refresh_token&refresh_token=${this.getRefreshToken()}`;
+    refreshUrl += `&client_id=${clientId}&client_secret=${clientSecret}`;
+    // const body = {
+    //   grant_type: 'refresh_token',
+    //   refresh_token: this.getRefreshToken(),
+    //   client_id: clientId,
+    //   client_secret: clientSecret,
+    // };
+
+    const fetchOptions = {
+      method: 'POST',
+      //body: JSON.stringify(body),
```

Comment:
nit: Can we add the headers in this object construction as well? Seems unnecessary to split that up.

Reply:
Weirdly, this throws linting errors. I can work on this at a later point when I clean up the SDKs